accepted/00080-api-versioning.md

```diff
@@ -0,0 +1,295 @@
+- Feature Name: Improve the 3rd party usage of XMLRPC API
+- Start Date: 2021-05-20
+- RFC PR:
+
+# Summary and motivation
+
+The primary goal of this RFC is to highlight the problems of writing
+client applications or scripts compatible with multiple versions of
+Uynui and SUSE Manager (SUMA) API and suggest an improvement in this
+area.
+
+When the consumers use the API, there must be a way for them to make
+sure they call the endpoints correctly (they need to call an existing
+method with correct parameters and they also need to make some
+assumptions on the return value structure).
+
+TODO keep here? :
+The secondary goal is to provide better guidelines for writing scripts
+that consume the API in the `API > FAQ` section of the Uyuni/SUMA Web
+UI or in the documentation.
+
+
+## XMLRPC compatibility
+
+Uyuni/SUMA exposes the information about the API and its structure in
+3 places:
+
+1. Information about the versions via the `api` XMLRPC endpoint:
+   - `getVersion`: the API version, e.g. `25`. This integer grows with
+   time, but there are no strict rules about it.
+   - `systemVersion`: the Uyuni server version (e.g. `4.2.0 Beta1` for
+   SUMA, `2021.05` for Uyuni)
+
+2. Furthermore, the `api` namespace also exposes basic API
+   introspection calls describing the structure of the call:
+   - `getApiNamespaces`: all API namespaces
+   - `getApiNamespaceCallList`: API methods of given namespace
+     together with their parameter types, faults and return types
+   - `getApiCallList`: all API methods grouped by namespace
+
+3. Faults (retrospective): When the API consumers call a method in a
+   wrong way (incorrect parameters, missing method or namespace), they
+   get back an XMLRPC fault with code `-1`.
+   For instance:
+   ```python
+   # Non-existing method/wrong method signature
+   <Fault -1: 'redstone.xmlrpc.XmlRpcFault: Could not find method:
+   test in class: com.redhat.rhn.frontend.xmlrpc.ansible.AnsibleHandler with params: []'>
+
+   # Non-existing handler
+   <Fault -1: 'The specified handler cannot be found'>
+   ```
+   Note: This fault code is
+   not reserved solely to report the API calling mismatch cases, but
+   is used in other cases (see the`CustomInfoHandler.java` file in
+   Uyuni).
```

Comment:
I am inclined to call this a bug. XMLRPC fault codes are meant to be unique.

Of course, this would be an API-compatibility-breaking bug :-(

Reply:
Same opinion here.

> Of course, this would be an API-compatibility-breaking bug :-(

We can fix it and bump the api version :wink: